packages/svelte/src/internal/client/proxy/proxy.js

```diff
@@ -0,0 +1,169 @@
+import {
+	effect_active,
+	get,
+	set,
+	increment,
+	source,
+	updating_derived,
+	UNINITIALIZED
+} from '../runtime.js';
+import { define_property, get_descriptor, is_array } from '../utils.js';
+import { READONLY_SYMBOL } from './readonly.js';
+
+/** @typedef {{ s: Map<string | symbol, import('../types.js').SourceSignal<any>>; v: import('../types.js').SourceSignal<number>; a: boolean }} Metadata */
+/** @typedef {Record<string | symbol, any> & { [STATE_SYMBOL]: Metadata }} StateObject */
+
+export const STATE_SYMBOL = Symbol('$state');
+
+const object_prototype = Object.prototype;
+const array_prototype = Array.prototype;
+const get_prototype_of = Object.getPrototypeOf;
+const is_frozen = Object.isFrozen;
+
+/**
+ * @template {StateObject} T
+ * @param {T} value
+ * @returns {T}
+ */
+export function proxy(value) {
+	if (typeof value === 'object' && value != null && !is_frozen(value) && !(STATE_SYMBOL in value)) {
+		const prototype = get_prototype_of(value);
+
+		// TODO handle Map and Set as well
+		if (prototype === object_prototype || prototype === array_prototype) {
+			define_property(value, STATE_SYMBOL, { value: init(value), writable: false });
```

Comment:
Can we get a test doing `for…in` on the original object to ensure we don't get the Symbol now? I imagine we might regress this unless we do in the future.

Reply:
Ah yeah ignore me then.